blocks/certifier.go

```diff
@@ -203,27 +214,50 @@ func (c *Certifier) RegisterForCert(ctx context.Context, lid types.LayerID, bid
 
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	c.createIfNeeded(lid, bid)
-	c.certifyMsgs[lid][bid].registered = true
-	return c.tryGenCert(ctx, logger, lid, bid)
+	info := c.createIfNeeded(lid, bid)
+	info.registered = true
+	return c.tryGenCert(ctx, logger, lid, bid, info)
 }
 
 // CertifyIfEligible signs the hare output, along with its role proof as a certifier, and gossip the CertifyMessage
 // if the node is eligible to be a certifier.
-func (c *Certifier) CertifyIfEligible(ctx context.Context, logger log.Log, lid types.LayerID, bid types.BlockID) error {
+func (c *Certifier) CertifyIfEligible(ctx context.Context, lid types.LayerID, bid types.BlockID) error {
 	if _, err := c.beacon.GetBeacon(lid.GetEpoch()); err != nil {
 		return errBeaconNotAvailable
 	}
-	// check if the node is eligible to certify the hare output
-	proof, err := c.oracle.Proof(ctx, lid, eligibility.CertifyRound)
+
+	var (
+		errsMu sync.Mutex
+		errs   error
+		eg     errgroup.Group
+	)
+	c.mu.Lock()
+	for _, s := range c.signers {
+		s := s
+		eg.Go(func() error {
```

Comment:
btw parallelization here is not an obvious choice.. there are two expensive operations (ed25519 and ecvrf) and both are measured in us, so maybe single key is 400us. parallelization is maybe useful for > 1000 signers, but then no need to create 1000 goroutines and it should be limited by number of cores